lib/netaddrq.c

```diff
@@ -0,0 +1,723 @@
+				UL_DEBUG_MASKNAMES(netaddrq)));
+}
+
+static inline enum ul_netaddrq_ip_rating
+evaluate_ip_quality(struct ul_nl_addr *addr) {
+	enum ul_netaddrq_ip_rating quality;
+
+	switch (addr->ifa_scope) {
+	case RT_SCOPE_UNIVERSE:
+		quality = ULNETLINK_RATING_SCOPE_UNIVERSE;
+		break;
+	case RT_SCOPE_LINK:
+		quality = ULNETLINK_RATING_SCOPE_LINK;
+		break;
+	case RT_SCOPE_SITE:
+		quality = ULNETLINK_RATING_SCOPE_SITE;
+		break;
+	default:
+		quality = ULNETLINK_RATING_BAD;
+		break;
+	}
+	if (addr->ifa_flags & IFA_F_TEMPORARY) {
+		if (quality <= ULNETLINK_RATING_F_TEMPORARY)
+			quality = ULNETLINK_RATING_F_TEMPORARY;
+	}
+	return quality;
+}
+
+#define DBG_CASE(x) case x: str = #x; break
+#define DBG_CASE_DEF8(x) default: snprintf(strx+2, 3, "%02hhx", x); str = strx; break
+static char *ip_rating_as_string(enum ul_netaddrq_ip_rating q)
+{
+	char *str;
+	static char strx[5] = "0x";
+	switch (q) {
+		DBG_CASE(ULNETLINK_RATING_SCOPE_UNIVERSE);
+		DBG_CASE(ULNETLINK_RATING_SCOPE_SITE);
+		DBG_CASE(ULNETLINK_RATING_F_TEMPORARY);
+		DBG_CASE(ULNETLINK_RATING_SCOPE_LINK);
+		DBG_CASE(ULNETLINK_RATING_BAD);
+		DBG_CASE_DEF8(q);
+	}
+	return str;
+}
+
+/* Netlink callback evaluating the address quality and building the list of
+ * interface lists */
+static int callback_addrq(struct ul_nl_data *nl) {
+	struct ul_netaddrq_data *addrq = UL_NETADDRQ_DATA(nl);
+	struct list_head *li, *ipq_list;
+	struct ul_netaddrq_iface *ifaceq = NULL;
+	struct ul_netaddrq_ip *ipq = NULL;
+	int rc;
+	bool *ifaces_change;
+
+	DBG(LIST, ul_debugobj(addrq, "callback_addrq() for %s on %s",
+			      ul_nl_addr_ntop_address(&(nl->addr)),
+			      nl->addr.ifname));
+	if (addrq->callback_pre)
+	{
+		DBG(LIST, ul_debugobj(addrq, "callback_pre"));
+		if ((rc = (*(addrq->callback_pre))(nl)))
+			DBG(LIST, ul_debugobj(nl, "callback_pre rc != 0"));
+	}
+
+	/* Search for interface in ifaces */
+	addrq->nifaces = 0;
+
+	list_for_each(li, &(addrq->ifaces)) {
+		struct ul_netaddrq_iface *ifaceqq;
+
+		ifaceqq = list_entry(li, struct ul_netaddrq_iface, entry);
+		if (ifaceqq->ifa_index == nl->addr.ifa_index) {
+			ifaceq = ifaceqq;
+			DBG(LIST, ul_debugobj(ifaceq,
+					      "%s found in addrq",
+					      nl->addr.ifname));
+			break;
+		}
+		addrq->nifaces++;
+	}
+
+	if (ifaceq == NULL) {
+		if (nl->rtm_event) {
+			if (addrq->nifaces >= max_ifaces) {
+				DBG(LIST, ul_debugobj(addrq,
+						       "too many interfaces"));
+				addrq->overflow = true;
+				return UL_NL_IFACES_MAX;
+			}
+			DBG(LIST, ul_debugobj(addrq,
+					       "new ifa_index in addrq"));
+			if (!(ifaceq = malloc(sizeof(struct ul_netaddrq_iface))))
+				return -ENOMEM;
+			INIT_LIST_HEAD(&(ifaceq->ip_quality_list_4));
+			INIT_LIST_HEAD(&(ifaceq->ip_quality_list_6));
+			ifaceq->ifa_index = nl->addr.ifa_index;
+			if (!(ifaceq->ifname = strdup(nl->addr.ifname)))
+			{
+				DBG(LIST, ul_debugobj(addrq,
+						       "malloc() failed"));
+				free(ifaceq);
+				return -1;
```

Comment:
Here should be -ENOMEM rather than -1 to be consistent.

Reply:
Fixed.